src/main/java/io/atomix/client/PrimitiveBuilder.java

```diff
@@ -0,0 +1,137 @@
+// SPDX-FileCopyrightText: 2022-present Intel Corporation
+//
+// SPDX-License-Identifier: Apache-2.0
+
+package io.atomix.client;
+
+
+import io.grpc.Channel;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionException;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+/**
+ * Abstract builder for distributed primitives.
+ *
+ * @param <B> builder type
+ * @param <P> primitive type
+ */
+public abstract class PrimitiveBuilder<B extends PrimitiveBuilder<B, P>, P extends SyncPrimitive> {
+    protected final String name;
+    protected final Channel channel;
+    protected boolean readOnly;
+    protected Map<String, String> tags = new HashMap<>();
+
+    protected PrimitiveBuilder(String name, Channel channel) {
+        this.name = checkNotNull(name, "primitive name cannot be null");
+        this.channel = checkNotNull(channel, "primitive channel cannot be null");
+    }
+
+    /**
+     * Returns the primitive name.
+     *
+     * @return the primitive name
+     */
+    protected String name() {
+        return name;
+    }
+
+    /**
+     * Returns the service channel.
+     *
+     * @return the service channel
+     */
+    protected Channel channel() {
+        return channel;
+    }
+
+    /**
+     * Returns the primitive tags.
+     *
+     * @return the primitive tags
+     */
+    protected Map<String, String> tags() {
```

Comment:
Suggest to improve documentation. Are `tags` in `PrimitiveBuilder` k8s labels ? Maybe should clarified in the comments what refers to and what we can influence with them ?